### PR TITLE
Add the TypeScript code and docs for Chapter of Hash Map #113

### PR DESCRIPTION
**Pull request by danielsss**
>  Added typescript `Hash Map`

### Checklist

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be run on a standard system (Win, MacOS, Ubuntu).

### CODE REVIEW

File: codes/typescript/tsconfig.json
Comment (krahets) [inline codes/typescript/tsconfig.json:1]: Is this config file necessary?

File: codes/typescript/chapter_hashing/array_hash_map.ts
Comment (krahets) [inline codes/typescript/chapter_hashing/array_hash_map.ts:18]: Could we use the built-in list `[]` instead?  You don't have to implement the ArrayList in Java. `[]` is the ArrayList in JS and TS.

File: codes/typescript/chapter_hashing/array_hash_map.ts
Comment (krahets) [inline codes/typescript/chapter_hashing/array_hash_map.ts:139]: `export {};` Please follow the formating of the previous TS code.

File: codes/typescript/chapter_hashing/array_hash_map.ts
Comment (krahets) [inline codes/typescript/chapter_hashing/array_hash_map.ts:9]: The indentation is 4 spaces for this project. Please compare the code with the previous TS code and match the formmating.

**Review by krahets — COMMENTED** [review]
Thank @danielsss for the contribution! Please address the comments.

**Comment by danielsss** [comment]
> Thank @danielsss for the contribution! Please address the comments.

Hi @krahets, All the issues that you mentioned above have been addressed.

File: codes/typescript/tsconfig.json
Comment (danielsss) [reply codes/typescript/tsconfig.json]: I was thinking to add some proper test cases into the `typescript`, but it's gonna be fine to remove it.

Comment (krahets) [reply codes/typescript/tsconfig.json]: What is the relationship between the test case and this config file?

File: codes/typescript/chapter_hashing/array_hash_map.ts
Comment (krahets) [inline codes/typescript/chapter_hashing/array_hash_map.ts:97]: Add `/* Driver Code */` here.

File: docs/chapter_hashing/hash_map.md
Comment (krahets) [inline docs/chapter_hashing/hash_map.md:257]: Same as above.

File: docs/chapter_hashing/hash_map.md
Comment (krahets) [inline docs/chapter_hashing/hash_map.md:558]: Is it necessary to write as any here, and why?

**Review by krahets — COMMENTED** [review]
Thanks for the update! Several comments are to be addressed. Additionally, the `hash_map.ts` file is lost, which is designed to introduce the built-in HashMap.

**Comment by danielsss** [comment]
> Thanks for the update! Several comments are to be addressed. Additionally, the `hash_map.ts` file is lost, which is designed to introduce the built-in HashMap.

Thanks for your review. The issues are resolved.

**Review by krahets — COMMENTED** [review]
Thanks for the changes. Waiting for @justin-tse for the final review.

File: codes/typescript/chapter_hashing/hash_map.ts
Comment (justin-tse) [inline codes/typescript/chapter_hashing/hash_map.ts:10]: Your class is not executed. You should test the built-in Map method, so you do not need to use class.

**Review by justin-tse — CHANGES_REQUESTED** [review]
Thank @danielsss for the contribution! I have downloaded your code and test but I found your array_hash_map and hash_map codes could not pass the test. Please make sure your codes pass the test and ensure the outputs are the same as the outputs of reference codes(Java codes). Thank you!

File: codes/typescript/chapter_hashing/hash_map.ts
Comment (danielsss) [reply codes/typescript/chapter_hashing/hash_map.ts]: I've removed the `hash_map` class and only test codes are left.

Although it's not totally identical to the reference codes of Java, but definitely enough to convey the circumstances. The codes have also been tested by the below commands(I don't know what kind of test method you used, but `deno` can be easily run as `Typescript` doing so):

```shell
deno run codes/typescript/chapter_hashing/hash_map.ts
添加完成后，哈希表为
Key -> Value
Map { 12836 => "小哈", 15937 => "小啰", 16750 => "小算", 13276 => "小法", 10583 => "小鸭" }

输入学号 15937 ，查询到姓名 小啰

删除 10583 后，哈希表为
Key -> Value
Map { 12836 => "小哈", 15937 => "小啰", 16750 => "小算", 13276 => "小法" }

遍历键值对 Key->Value
12836 -> 小哈
15937 -> 小啰
16750 -> 小算
13276 -> 小法

单独遍历键 Key
12836
15937
16750
13276

单独遍历值 Value
小哈
小啰
小算
小法
```

```shell
deno run codes/typescript/chapter_hashing/array_hash_map.ts
添加完成后，哈希表为
Key -> Value
36 -> 小哈
37 -> 小啰
50 -> 小算
76 -> 小法
83 -> 小鸭

输入学号 15937 ，查询到姓名 小啰

删除 10583 后，哈希表为
Key -> Value
36 -> 小哈
37 -> 小啰
50 -> 小算
76 -> 小法

遍历键值对 Key->Value
36 -> 小哈
37 -> 小啰
50 -> 小算
76 -> 小法

单独遍历键 Key
36
37
50
76

单独遍历值 Value
小哈
小啰
小算
小法

```

Comment (justin-tse) [reply codes/typescript/chapter_hashing/hash_map.ts]: Thank you! @danielsss I see you are using deno to test but you should know that deno has skipped type-checking by default. You could see the article: https://github.com/denoland/deno/issues/11340
If you use deno to test, you need to use the --check option (deno run --check  xxxxxx.ts). Or you could use ts-node to test.

Comment (justin-tse) [reply codes/typescript/chapter_hashing/hash_map.ts]: Thank you! @danielsss I see you are using deno to test but you should know that deno has skipped type-checking by default. You could see the article: https://github.com/denoland/deno/issues/11340
If you use deno to test, you need to use the --check option 
```
 deno run --check  xxxxxx.ts
```
Or you could use ts-node to test. And then you will see the type error message, please make sure you could pass it.

**Review by justin-tse — COMMENTED** [review]
Thank you! @danielsss I see you are using deno to test but you should know that deno has skipped type-checking by default. You could see the article: https://github.com/denoland/deno/issues/11340
If you use deno to test, you need to use the --check option 
```
 deno run --check  xxxxxx.ts
```
Or you could use ts-node to test. And then you will see the type error message, please make sure you could pass it.

**Comment by danielsss** [comment]
> Thank you! @danielsss I see you are using deno to test but you should know that deno has skipped type-checking by default. You could see the article: [denoland/deno#11340](https://github.com/denoland/deno/issues/11340) If you use deno to test, you need to use the --check option
> 
> ```
>  deno run --check  xxxxxx.ts
> ```
> 
> Or you could use ts-node to test. And then you will see the type error message, please make sure you could pass it.

All the type checks have been resolved, and now it can be passed by `--check`.

**Review by justin-tse — CHANGES_REQUESTED** [review]
Thank you @danielsss  for your clear code. It seems good and only has a small flaw.

**Review by krahets — APPROVED** [review]
Now merge it.

**Pull request merged**
